File: sirius-core/src/main/java/org/starcoin/sirius/core/Eon.java
package org.starcoin.sirius.core;

public final class Eon {

    public int id;
    public Epoch epoch;

    public Eon(int id, Epoch epoch) {
        this.id = id;
        this.epoch = epoch;
    }

    public Eon() {
    }

    public enum Epoch {
        FIRST,
        SECOND,
        THIRD,
        LAST;
    }

    public static Eon calculateEon(int blockHeight, int blocksPerEon) {
        return new Eon(blockHeight / blocksPerEon,
                Epoch.values()[(blockHeight % blocksPerEon) / (blocksPerEon / 4)]);
    }
}

File: sirius-core/src/main/kotlin/org/starcoin/sirius/core/Eon.kt
package org.starcoin.sirius.core

class Eon(val id: Int, val epoch: Epoch) {

    constructor(id: Long, epoch: Epoch) : this(id.toInt(), epoch)

    enum class Epoch {
        FIRST,
        SECOND,
        THIRD,
        LAST
    }

    companion object {

        fun calculateEon(startBlockNumber: Long, currentBlockNumber: Long, blocksPerEon: Int): Eon {
            val blocks = (currentBlockNumber - startBlockNumber).toInt()
            return Eon(
                blocks / blocksPerEon,
                Epoch.values()[(blocks % blocksPerEon / (blocksPerEon / 4))]
            )
        }

        fun waitToEon(
            startBlockNumber: Long,
            currentBlockNumber: Long,
            blocksPerEon: Int,
            expectEon: Int
        ): Int {
            return blocksPerEon * expectEon - (currentBlockNumber - startBlockNumber).toInt()
        }

        fun waitBlockNumber(
            startBlockNumber: Long,
            blocksPerEon: Int,
            expectEon: Int
        ): Long {
            return startBlockNumber + blocksPerEon * expectEon
        }
    }
}
